File: services/shared/src/events.ts
export interface EventEnvelope<T> {
  event_id: string;
  timestamp: string;
  payload: T;
}

// User events
export interface UserCreated {
  user_id: string;
  username: string;
  email: string;
}

export interface UserUpdated {
  user_id: string;
  username?: string;
  bio?: string;
}

export interface UserDeleted {
  user_id: string;
}

// Post events
export interface PostCreated {
  post_id: string;
  author_id: string;
  title: string;
  slug: string;
  content: string;
  tags: string[];
  status: string;
}

export interface PostUpdated {
  post_id: string;
  title: string;
  slug: string;
  content: string;
  tags: string[];
  status: string;
}

export interface PostPublished {
  post_id: string;
  author_id: string;
  title: string;
  slug: string;
}

export interface PostDeleted {
  post_id: string;
}

// Comment events
export interface CommentCreated {
  comment_id: string;
  post_id: string;
  author_id: string;
  content: string;
  parent_id: string | null;
}

export interface CommentDeleted {
  comment_id: string;
  post_id: string;
}

// Media events
export interface MediaUploaded {
  media_id: string;
  user_id: string;
  filename: string;
  content_type: string;
  size: number;
}

export interface MediaDeleted {
  media_id: string;
}

export function createEnvelope<T>(payload: T): EventEnvelope<T> {
  return {
    event_id: crypto.randomUUID(),
    timestamp: new Date().toISOString(),
    payload,
  };
}

File: crates/shared/src/events.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EventEnvelope<T: Serialize> {
    pub event_id: Uuid,
    pub timestamp: DateTime<Utc>,
    pub payload: T,
}

impl<T: Serialize> EventEnvelope<T> {
    pub fn new(payload: T) -> Self {
        Self {
            event_id: Uuid::new_v4(),
            timestamp: Utc::now(),
            payload,
        }
    }
}

// User events
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserCreated {
    pub user_id: Uuid,
    pub username: String,
    pub email: String,
    #[serde(default)]
    pub role: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserUpdated {
    pub user_id: Uuid,
    pub username: Option<String>,
    #[serde(default)]
    pub email: Option<String>,
    pub bio: Option<String>,
    #[serde(default)]
    pub role: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserDeleted {
    pub user_id: Uuid,
}

// Post events
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PostCreated {
    pub post_id: Uuid,
    pub author_id: Uuid,
    pub title: String,
    pub slug: String,
    pub content: String,
    pub tags: Vec<String>,
    pub status: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PostUpdated {
    pub post_id: Uuid,
    pub title: String,
    pub slug: String,
    pub content: String,
    pub tags: Vec<String>,
    pub status: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PostPublished {
    pub post_id: Uuid,
    pub author_id: Uuid,
    pub title: String,
    pub slug: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PostDeleted {
    pub post_id: Uuid,
}

// Comment events
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CommentCreated {
    pub comment_id: Uuid,
    pub post_id: Uuid,
    pub author_id: Uuid,
    pub content: String,
    pub parent_id: Option<Uuid>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CommentDeleted {
    pub comment_id: Uuid,
    pub post_id: Uuid,
}

// Media events
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MediaUploaded {
    pub media_id: Uuid,
    pub user_id: Uuid,
    pub filename: String,
    pub content_type: String,
    pub size: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MediaDeleted {
    pub media_id: Uuid,
}
